frontend/catalyst/utils/wrapper.cpp

```diff
@@ -1,15 +1,186 @@
+#include <iostream>
 #include <pybind11/pybind11.h>
 
+#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
+#define NUMPY_WARN_IF_NO_MEM_POLICY 1
+#include "numpy/arrayobject.h"
+
 namespace py = pybind11;
 
-void wrap(py::object func, py::tuple py_args)
+struct memref_beginning_t {
+    char *allocated;
+    char *aligned;
+    size_t offset;
+};
+
+size_t memref_size_based_on_rank(size_t rank)
+{
+    size_t allocated = sizeof(void *);
+    size_t aligned = sizeof(void *);
+    size_t offset = sizeof(size_t);
+    size_t sizes = rank * sizeof(size_t);
+    size_t strides = rank * sizeof(size_t);
+    return allocated + aligned + offset + sizes + strides;
+}
+
+size_t *to_sizes(char *base, size_t rank)
+{
+    if (rank == 0) {
+        return NULL;
+    }
+
+    size_t allocated = sizeof(void *);
+    size_t aligned = sizeof(void *);
+    size_t offset = sizeof(size_t);
+    size_t bytes_offset = allocated + aligned + offset;
+    return (size_t *)(base + bytes_offset);
+}
+
+size_t *to_strides(char *base, size_t rank)
+{
+    if (rank == 0) {
+        return NULL;
+    }
+
+    size_t allocated = sizeof(void *);
+    size_t aligned = sizeof(void *);
+    size_t offset = sizeof(size_t);
+    size_t sizes = rank * sizeof(size_t);
+    size_t bytes_offset = allocated + aligned + offset + sizes;
+    return (size_t *)(base + bytes_offset);
+}
+
+void free_wrap(PyObject *capsule)
+{
+    void *obj = PyCapsule_GetPointer(capsule, NULL);
+    void *allocated = obj;
+    free(allocated);
```

Comment:
What does the middle copy do?

Reply:
I'll delete it. Remnants from debugging.